### PR TITLE
[3/N][2D] Enable training with new 2D flow

### PR DESCRIPTION
**Pull request by wz337**
This PR enables training with new 2D flow and adds associated test. 

state_dict related changes would be in later PRs. 

cc. @fegin, @fduwjj, @wanchaol, @awgu

### CODE REVIEW

**Comment by wz337** [comment]
Unrelated test failure. https://github.com/pytorch/pytorch/actions/runs/6230762730/job/16911903503?pr=109553

File: torch/distributed/fsdp/_init_utils.py [hunk 1]
```diff
@@ -506,6 +506,30 @@ def _init_prefetching_state(
...
+def _init_extension(state: _FSDPState, device_mesh: DeviceMesh = None) -> _FSDPState:
+    # TODO: we need to add additional check once we support FSDP + PiPPy.
+    # This check is currently sufficient, since we only support FSDP + TP.
+    if device_mesh:
+        state._enable_extension = (
+            mesh_resources.get_parent_mesh(state._device_mesh) is not None
+        )
+
+        if state._enable_extension:
+            try:
```
Comment (fegin) [inline torch/distributed/fsdp/_init_utils.py:519]: We don't need this try/except for the import, right? There is no way `_init_extension` exist but those extensions are not in the code base.

File: torch/distributed/tensor/parallel/fsdp.py [hunk 2]
```diff
@@ -3,19 +3,48 @@
...
     _flatten_tensor,
     _pre_load_state_dict,
     _unflatten_tensor,
 )
 
-__all__ = ["enable_2d_with_fsdp"]
+__all__ = ["enable_2d_with_fsdp", "DTensorExtensions"]
+
+
+class DTensorExtensions(FSDPExtensions):
```
Comment (fegin) [inline torch/distributed/tensor/parallel/fsdp.py:18]: Can we add a one line docstring to explicitly mention this is used for 2D?

**Review by fegin — APPROVED** [review]

**Comment by wz337** [comment]
@awgu Tagging you as we are introducing changes in the FSDP __init__().

File: test/distributed/tensor/parallel/test_fsdp_2d_parallel.py [hunk 3]
```diff
@@ -145,6 +145,15 @@ def _apply_optim_in_backward(param_group):
     )
 
 
+def _clean_up_fsdp_param_name(name):
```
Comment (awgu) [inline test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:148]: Curious: Is this different from `clean_tensor_name()`?
https://github.com/pytorch/pytorch/blob/11c6a98bcae9e5ae305f8b21b9180a5ee2e1ba71/torch/distributed/fsdp/_common_utils.py#L228

**Review by awgu — APPROVED** [review]

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -491,6 +492,8 @@ def __init__(
         _init_runtime_state(self)
         _init_prefetching_state(self, backward_prefetch, forward_prefetch)
         _init_buffer_state(self, module)
+        # extension needs to be set before `_init_param_handle_from_module()`
+        _init_extension(self, device_mesh)
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:496]: For completeness, shall we include this change for `fully_shard` path too (even though it is not too important right now)?
https://github.com/pytorch/pytorch/blob/c2f5d4d8f09b535e434313223434009bad282258/torch/distributed/_composable/fully_shard.py#L106-L109

File: test/distributed/tensor/parallel/test_fsdp_2d_parallel.py (hunk 3 above)
Comment (wz337) [reply test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:148]: Thanks! I was not aware of this in the _common_utils. Will re-use then!

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 4 above)
Comment (wz337) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:496]: > For completeness, shall we include this change for `fully_shard` path too (even though it is not too important right now)?
> 
> https://github.com/pytorch/pytorch/blob/c2f5d4d8f09b535e434313223434009bad282258/torch/distributed/_composable/fully_shard.py#L106-L109

Yes. Good point! We are planning to do the composable API changes in follow up PRs.

**Comment by wz337** [comment]
pytorchmergebot rebase -b main

File: test/distributed/tensor/parallel/test_fsdp_2d_parallel.py [hunk 5]
```diff
@@ -314,8 +313,25 @@ def test_2d_fsdp_integration_fsdp_nested_param_groups(self) -> None:
 
 
 class TestNew2dParallelIntegration(DTensorTestBase):
+    # TODO: this is duplicate code from above, but once we remove the enable_2d_with_fsdp(),
+    # we will remove the above test class Test2dParallelIntegration.
+    def _compare_params(self, m1, m2):
```
Comment (fduwjj) [inline test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:318]: Can we reuse the func in L170?

File: test/distributed/tensor/parallel/test_fsdp_2d_parallel.py [hunk 6]
```diff
@@ -327,6 +343,78 @@ def test_2d_fsdp_state_enable_extension(self):
         fsdp_state = _get_module_fsdp_state(model)
         self.assertEqual(fsdp_state._enable_extension, True)
 
+    def _test_2d_e2e_training(
```
Comment (fduwjj) [inline test/distributed/tensor/parallel/test_fsdp_2d_parallel.py:346]: Can we consolidate with L194? Feel free to refactor the code as you feel necessary.

**Pull request closed without merging**